server: TestDrain failed

server.TestDrain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21226939?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21226939?buildTab=artifacts#/) on master @ [bb02fd678b06df257d9296b22ceb48cba45f0693](https://github.com/cockroachdb/cockroach/commits/bb02fd678b06df257d9296b22ceb48cba45f0693):


```
=== RUN   TestDrain
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/4b992d1f969871bcc9fd338a5653884d/logTestDrain3341812692
    test_log_scope.go:82: use -show-logs to present logs inline
--- FAIL: TestDrain (4.48s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1b7e3e0]

goroutine 17366 [running]:
testing.tRunner.func1.2({0x5443000, 0xabc2b40})
	GOROOT/src/testing/testing.go:1872 +0x190
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1875 +0x31c
panic({0x5443000?, 0xabc2b40?})
	GOROOT/src/runtime/panic.go:783 +0x120
github.com/cockroachdb/cockroach/pkg/util/grpcutil.IsClosedConnection({0x0, 0x0})
	pkg/util/grpcutil/grpc_util.go:94 +0xf0
github.com/cockroachdb/cockroach/pkg/server_test.doTestDrain.func3()
	pkg/server/drain_test.go:130 +0x6c
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1()
	pkg/testutils/soon.go:95 +0x38
github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration(0xa7a358200, 0x400375fd30)
	pkg/util/retry/retry.go:362 +0xbc
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError(0x400359fec0, 0xa7a358200)
	pkg/testutils/soon.go:101 +0x4c
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x6dbf030, 0x4001cff2f0}, 0x400359fec0, 0xa7a358200)
	pkg/testutils/soon.go:63 +0x44
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x6dbf030, 0x4001cff2f0}, 0x400359fec0)
	pkg/testutils/soon.go:42 +0x4c
github.com/cockroachdb/cockroach/pkg/server_test.doTestDrain(0x4011d5ac40, 0x0)
	pkg/server/drain_test.go:128 +0x304
github.com/cockroachdb/cockroach/pkg/server_test.TestDrain.func1(0x4016bbef58?, 0xac?)
	pkg/server/drain_test.go:40 +0x1c
github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
	pkg/testutils/subtest.go:23 +0x28
testing.tRunner(0x4011d5ac40, 0x401dc290e0)
	GOROOT/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 17361
	GOROOT/src/testing/testing.go:1997 +0x364
=== RUN   TestDrain/secondaryTenants=false
    test_server_shim.go:638: DRPC is enabled (via testserver factory defaults)
    drain_test.go:105: drain response: is_draining:true 
--- FAIL: TestDrain/secondaryTenants=false (4.48s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-61358

server.TestDrain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21250960?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21250960?buildTab=artifacts#/) on master @ [86c176dbb4e93eda87764ced207d046e81057e14](https://github.com/cockroachdb/cockroach/commits/86c176dbb4e93eda87764ced207d046e81057e14):


```
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError(0x4010821ec0, 0xa7a358200)
	pkg/testutils/soon.go:82 +0x4c
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x6ead550, 0x401cdf9c50}, 0x4010821ec0, 0xa7a358200)
	pkg/testutils/soon.go:59 +0x44
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x6ead550, 0x401cdf9c50}, 0x4010821ec0)
	pkg/testutils/soon.go:38 +0x4c
github.com/cockroachdb/cockroach/pkg/server_test.doTestDrain(0x4003d7fdc0, 0x1)
	pkg/server/drain_test.go:128 +0x304
github.com/cockroachdb/cockroach/pkg/server_test.TestDrain.func1(0x4019320f58?, 0xfc?)
	pkg/server/drain_test.go:40 +0x1c
github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
	pkg/testutils/subtest.go:23 +0x28
testing.tRunner(0x4003d7fdc0, 0x401e370000)
	GOROOT/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 16439
	GOROOT/src/testing/testing.go:1997 +0x364
=== RUN   TestDrain/secondaryTenants=true
    test_server_shim.go:638: DRPC is enabled (via testserver factory defaults)
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:115 drain_remaining_description:"descriptor leases: 30, liveness record: 1, range lease iterations: 83, SQL clients: 1" 
    drain_test.go:105: drain response: is_draining:true 
--- FAIL: TestDrain/secondaryTenants=true (5.64s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

server.TestDrain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21275261?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21275261?buildTab=artifacts#/) on master @ [33998e0a929b840bd679ca3a77b41eee49bf9f50](https://github.com/cockroachdb/cockroach/commits/33998e0a929b840bd679ca3a77b41eee49bf9f50):


```
	pkg/util/retry/retry.go:362 +0xbc
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError(0x401faa1ec0, 0xa7a358200)
	pkg/testutils/soon.go:82 +0x4c
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x6ee1390, 0x40195f16e0}, 0x401faa1ec0, 0xa7a358200)
	pkg/testutils/soon.go:59 +0x44
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x6ee1390, 0x40195f16e0}, 0x401faa1ec0)
	pkg/testutils/soon.go:38 +0x4c
github.com/cockroachdb/cockroach/pkg/server_test.doTestDrain(0x400f0e81c0, 0x1)
	pkg/server/drain_test.go:128 +0x304
github.com/cockroachdb/cockroach/pkg/server_test.TestDrain.func1(0x4001083f58?, 0xfc?)
	pkg/server/drain_test.go:40 +0x1c
github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
	pkg/testutils/subtest.go:23 +0x28
testing.tRunner(0x400f0e81c0, 0x401a89e000)
	GOROOT/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 16552
	GOROOT/src/testing/testing.go:1997 +0x364
=== RUN   TestDrain/secondaryTenants=true
    test_server_shim.go:638: DRPC is enabled (via testserver factory defaults)
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:1 drain_remaining_description:"tenant servers: 1" 
    drain_test.go:105: drain response: is_draining:true drain_remaining_indicator:116 drain_remaining_description:"liveness record: 1, range lease iterations: 84, SQL clients: 1, descriptor leases: 30" 
    drain_test.go:105: drain response: is_draining:true 
--- FAIL: TestDrain/secondaryTenants=true (5.43s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/drain_test.go
+++ b/pkg/server/drain_test.go
@@ -127,7 +127,7 @@ func doTestDrain(tt *testing.T, secondaryTenants bool) {
 	// Now expect the server to be shut down.
 	testutils.SucceedsSoon(t, func() error {
 		_, err := t.c.Drain(context.Background(), &serverpb.DrainRequest{Shutdown: false})
-		if grpcutil.IsClosedConnection(err) {
+		if err != nil && grpcutil.IsClosedConnection(err) { //nolint:returnerrcheck
 			return nil
 		}
 		// It is incorrect to use errors.Wrap since that will result in a nil

--- a/pkg/util/grpcutil/grpc_util.go
+++ b/pkg/util/grpcutil/grpc_util.go
@@ -80,6 +80,9 @@ func IsContextCanceled(err error) bool {
 // IsClosedConnection returns true if err's Cause is an error produced by gRPC
 // on closed connections.
 func IsClosedConnection(err error) bool {
+	if err == nil {
+		return false
+	}
 	if errors.Is(err, ErrConnectionInterrupted) {
 		return true
 	}

